### PR TITLE
Support for Count Expressions

### PR DESCRIPTION
**Pull request by adamretter**
This is currently passing 12 out of 13 XQTS tests. The 1 failing test is: `prod-CountClause/count-009`. 

Screenshot 2023-06-06 at 23 20 43

## <a name="acceptance" id="acceptance">Acceptance of Failing XQTS Test</a>
The 1 failing test occurs due to the existing design and implementation of FLWOR expressions in eXist-db.

Specifically, eXist-db's implementation of the XQuery `order by` clause does not follow the W3C specification. The W3C XQuery 3.1 [Order By Clause](https://www.w3.org/TR/xquery-31/#id-order-by-clause) specification states:

> The purpose of an order by clause is to impose a value-based ordering on the tuples in the tuple stream. The output tuple stream of the order by clause contains the same tuples as its input tuple stream, but the tuples may be in a different order.

This means that any tuples in a tuple stream after an Order By Clause should be in the order defined by the order by clause. In the eXist-db implementation that is not the case. In eXist-db any expressions/clauses within a FLWOR expression that occur after the order by clause, but before the return clause, will not see an ordered tuple stream; instead they see the tuple stream prior to the order by clause. The reason for this is that in eXist-db the operation of executing the `sort` for the Order By clause is always deferred until just before the return clause of the FLWOR expression. In eXist-db this is known as a `postEval` phase, see:
1. [`FLWORClause#postEval(Sequence)`](https://github.com/eXist-db/exist/blob/eXist-6.2.0/exist-core/src/main/java/org/exist/xquery/FLWORClause.java#L95)
2. [`ForExpr#eval(Sequence, Item)`](https://github.com/eXist-db/exist/blob/eXist-6.2.0/exist-core/src/main/java/org/exist/xquery/ForExpr.java#L224)
3. [`OrderByClause#postEval(Sequence)`](https://github.com/eXist-db/exist/blob/eXist-6.2.0/exist-core/src/main/java/org/exist/xquery/OrderByClause.java#L90)

Unfortunately at this time it appears that modifying eXist-db to correctly implement the `Order By Clause` semantics would be a huge undertaking, and may require a rewrite of all FLWOR expressions. Therefore we consider it out of scope for this PR.

On an tangentially related note, we suspect there may be other problems with all other FLWOR expressions in eXist-db that implement `postEval` with regards to compliance with W3C XQuery 3.1.

----
### In Further Detail

Given the XQuery:
```xquery
for $x in ('a', 'b')

  count $index1
  
  let $remainder := $index1 mod 2
  order by $remainder, $index1
  
  count $index2

return
    <x index1="{$index1}" index2="{$index2}">{$x}</x>
```

The result should be computed as:
```xml
<x index1="2" index2="1">b</x>
<x index1="1" index2="2">a</x>
```

However eXist-db produces the *incorrect* result due to its incorrect implementation of the `order by` clause:
```xml
<x index1="2" index2="2">b</x>
<x index1="1" index2="1">a</x>
```

This occurs because eXist-db evaluates the query by effectively executing the following steps:
1. Bind `$x := 'a'`
2. Bind `$index1 := 1`
3. Bind `$remainder := 1`
4. Bind `$index2 := 1`
5. Add `<x index1="1" index2="1">a</x>` to the result sequence.
6. Bind `$x := 'b'`
7. Bind `$index1 := 2`
8. Bind `$remainder := 0`
9. Bind `$index2 := 2`
10. Add `<x index1="2" index2="2">b</x>` to the result sequence.
11. Sort the result sequence based by `$remainder ascending` and then `$index1 ascending`
12. Return the sorted result sequence.

The problem is that eXist-db applies the sorting (i.e. the `order by` clause) after the fact.

I have created a detailed UML Sequence Diagram of the above XQuery execution in eXist-db that shows the exact problem:
eXist-db Count and Order By - UML Sequence Diagram

In case it is helpful for anyone else, the code for the UML Sequence Diagram:
```uml
participant user as "User"
participant xquery as "XQuery"
participant pathexpr as "PathExpr"
participant x as "$x: ForExpr"
participant index1 as "$index1: CountExpr"
participant remainder as "$remainder: LetExpr"
participant orderby as "OrderByClause"
participant index2 as "$index2: CountExpr"
participant element as "ElementConstructor"

activate user

'XQuery Compilation Phase
note over user: XQuery Compilation phase starts here
activate xquery
user -> xquery: compile(...)
xquery -> xquery: analyzeAndOptimizeIfModulesChanged(...)
activate pathexpr
xquery -> pathexpr: analyze(...)
activate x
pathexpr -> x: analyze(...)
activate index1
x -> index1: analyze(...)
activate remainder
index1 -> remainder: analyze(...)
activate orderby
remainder -> orderby: analyze(...)
activate index2
orderby -> index2: analyze(...)
activate element
index2 -> element: analyze(...)
element -> index2
deactivate element
index2 -> orderby
deactivate index2
orderby -> remainder
deactivate orderby
remainder -> index1
deactivate remainder
index1 -> x
deactivate index1
x -> pathexpr
deactivate x
pathexpr -> xquery
deactivate pathexpr
xquery -> user
deactivate xquery

'XQuery Execution Phase
note over user: XQuery Execution phase starts here
activate xquery
user -> xquery: execute(...)
activate pathexpr
xquery -> pathexpr: eval(null)
activate x
pathexpr -> x: eval(null)
x -> x: isOuterFor == true
x -> x: $index1 is FLWORClause == true
activate index1
x -> index1: preEval(('a', 'b'))
index1 -> index1: hasPreviousOrderByDescending() == false
index1 -> index1: calcStartPosition == 0
activate remainder
index1 --> remainder: preEval(('a', 'b'))
activate orderby
remainder --> orderby: preEval(('a', 'b')) 
activate index2
orderby -> index2: preEval(('a', 'b')) 
index2 -> index2: hasPreviousOrderByDescending() == false
index2 -> index2: calcStartPosition == 0
index2 -> orderby
orderby --> remainder
remainder --> index1
index1 -> x

'Process first Item 'a'
x -> x: processItem('a')
x -> index1: eval(null, null)
index1 -> index1: increment/decrement count == 1
index1 -> remainder: eval(null, null)
remainder -> orderby: eval(null, null)
note over orderby, index2, element: Error: No ordering of the Tuple Stream as been performed by `order by` before\n the `count $index2` is evaluated and subsequently used by the `x` Element Constructor
orderby -> index2: eval(null, null)
index2 -> index2: increment/decrement count == 1
activate element
index2 -> element: eval(null, null)
element-> index2: (index1=1 index2=1 x='a')
index2-> orderby: (index1=1 index2=1 x='a')
orderby -> orderby: orderedResult == \n  [\n    (index1=1 index2=1 x='a')\n  ]
note over orderby: orderedResult is not yet sorted!
orderby -> remainder: (index1=1 index2=1 x='a')
remainder -> index1: (index1=1 index2=1 x='a')
index1 -> x: (index1=1 index2=1 x='a')

'Process second Item 'b'
x -> x: processItem('b')
x -> index1: eval(null, null)
index1 -> index1: increment/decrement count == 2
index1 -> remainder: eval(null, null)
note over orderby, index2, element: Error: No ordering of the Tuple Stream as been performed by `order by` before\n the `count $index2` is evaluated and subsequently used by the `x` Element Constructor
remainder -> orderby: eval(null, null)
orderby -> index2: eval(null, null)
index2 -> index2: increment/decrement count == 2
index2 -> element: eval(null, null)
element-> index2: (index1=2 index2=2 x='b')
deactivate element
index2-> orderby: (index1=2 index2=2 x='b')
orderby -> orderby: orderedResult == \n  [\n    (index1=1 index2=1 x='a')\n    (index1=2 index2=2 x='b')\n  ]
note over orderby: orderedResult is not yet sorted!
orderby -> remainder: (index1=2 index2=2 x='b')
remainder -> index1: (index1=2 index2=2 x='b')
index1 -> x: (index1=2 index2=2 x='b')

'Post Eval phase
x -> x: callPostEval == true
note over x: Post Eval Phase starts here...
x --> index1: postEval([\n    (index1=1 index2=1 x='a')\n    (index1=2 index2=2 x='b')\n  ])
index1 --> remainder: postEval([\n    (index1=1 index2=1 x='a')\n    (index1=2 index2=2 x='b')\n  ])
remainder --> orderby: postEval([\n    (index1=1 index2=1 x='a')\n    (index1=2 index2=2 x='b')\n  ])
orderby -> orderby: orderedResult.sort
orderby -> orderby: orderedResult == \n  [\n    (index1=2 index2=2 x='b')\n    (index1=1 index2=1 x='a')\n  ]
note over orderby: orderedResult is now sorted!
note over orderby: Error: The Tuple Stream was sorted after the `count $index2`,\n it should have been sorted before!
orderby -> index2: postEval([\n    (index1=2 index2=2 x='b')\n    (index1=1 index2=1 x='a')\n  ])
index2 -> orderby: [\n    (index1=2 index2=2 x='b')\n    (index1=1 index2=1 x='a')\n  ]
deactivate index2
orderby --> remainder: [\n    (index1=2 index2=2 x='b')\n    (index1=1 index2=1 x='a')\n  ]
deactivate orderby
remainder --> index1: [\n    (index1=2 index2=2 x='b')\n    (index1=1 index2=1 x='a')\n  ]
deactivate remainder
index1 --> x: [\n    (index1=2 index2=2 x='b')\n    (index1=1 index2=1 x='a')\n  ]
deactivate index1
x -> pathexpr: [\n    (index1=2 index2=2 x='b')\n    (index1=1 index2=1 x='a')\n  ]
deactivate x
pathexpr -> xquery: [\n    (index1=2 index2=2 x='b')\n    (index1=1 index2=1 x='a')\n  ]
deactivate pathexpr
note over user: The orderedResult overwrites the result sequence,\n and is returned to the User
xquery -> user: [\n    (index1=2 index2=2 x='b')\n    (index1=1 index2=1 x='a')\n  ]
deactivate xquery

deactivate user
```

----
This open source contribution to the eXist-db project was commissioned by the Office of the Historian, U.S. Department of State, https://history.state.gov/.

### CODE REVIEW

File: exist-core/src/main/java/org/exist/xquery/value/DoubleValue.java [hunk 1]
```diff
@@ -146,19 +146,25 @@ public boolean isPositive() {
...
+        } else if (isInfinite() && other.isInfinite() && isPositive() == other.isPositive()) {
+            comparison = () -> Constants.EQUAL;
+        } else if (other instanceof IntegerValue) {
+            comparison = () -> BigDecimal.valueOf(value).compareTo(new BigDecimal(((IntegerValue) other).value));
+        } else if (other instanceof DecimalValue) {
+            comparison = () -> BigDecimal.valueOf(value).compareTo(((DecimalValue) other).value);
+        } else if (other instanceof DoubleValue) {
+            comparison = () -> Double.compare(value, ((DoubleValue) other).value);
+        } else if (other instanceof FloatValue) {
+            comparison = () -> Double.compare(value, ((FloatValue) other).value);
```
Comment (reinhapa) [inline exist-core/src/main/java/org/exist/xquery/value/DoubleValue.java:163]: The enhanced `instanceof` with direct cast should be used here

File: exist-core/src/test/xquery/xquery3/serialize.xql [hunk 2]
```diff
@@ -254,7 +254,7 @@ function ser:serialize-empty-sequence() {
 
 declare
     %test:args(1234)
-    %test:assertEquals('1234')
+    %test:assertEquals(1234)
```
Comment (reinhapa) [inline exist-core/src/test/xquery/xquery3/serialize.xql:257]: Seems not related to the actual issue

File: exist-core/src/test/xquery/xquery3/serialize.xql [hunk 3]
```diff
@@ -265,7 +265,7 @@ function ser:adaptive-simple-atomic($atomic as xs:anyAtomicType) {
...
```
Comment (reinhapa) [inline exist-core/src/test/xquery/xquery3/serialize.xql:268]: Seems not related to the actual issue

**Review by reinhapa — CHANGES_REQUESTED** [review]

**Comment by dizzzz** [comment]
I missed the review request; I will have a look coming days.
Is there an easy way to seeing an list of review requests?

File: exist-core/src/main/java/org/exist/xquery/value/DoubleValue.java (hunk 1 above)
Comment (adamretter) [reply exist-core/src/main/java/org/exist/xquery/value/DoubleValue.java]: Done.

File: exist-core/src/test/xquery/xquery3/serialize.xql (hunk 2 above)
Comment (adamretter) [reply exist-core/src/test/xquery/xquery3/serialize.xql]: Removed.

File: exist-core/src/test/xquery/xquery3/serialize.xql (hunk 3 above)
Comment (adamretter) [reply exist-core/src/test/xquery/xquery3/serialize.xql]: Removed.

File: exist-core/src/main/java/org/exist/xquery/CountClause.java [hunk 4]
```diff
@@ -0,0 +1,211 @@
...
+        while (prev != null) {
+            switch (prev.getType()) {
+                case LET:
+                case GROUPBY:
+                case FOR:
+                    return false;
+                case ORDERBY:
+                   return isDescending(((OrderByClause) prev).getOrderSpecs());
+            }
+            prev = prev.getPreviousClause();
```
[the remark this answers is not in the document]
Comment (adamretter) [reply exist-core/src/main/java/org/exist/xquery/CountClause.java]: Done.

**Comment by adamretter** [comment]
> Is there an easy way to seeing an list of review requests?

@dizzzz Not that I know of. I think I normally get notifications via email myself.

File: exist-core/src/main/java/org/exist/xquery/CountClause.java [hunk 5]
```diff
@@ -0,0 +1,211 @@
...
+            context.getProfiler().end(this, "", resultSequence);
+        }
+
+        return resultSequence;
+    }
+
+    @Override
+    public Sequence postEval(Sequence seq) throws XPathException {
+        if (returnExpr instanceof FLWORClause) {
+            seq = ((FLWORClause)returnExpr).postEval(seq);
```
Comment (reinhapa) [inline exist-core/src/main/java/org/exist/xquery/CountClause.java:187]: Here the advanced `instanceof` could be used.

**Review by reinhapa — CHANGES_REQUESTED** [review]

File: exist-core/src/main/java/org/exist/xquery/CountClause.java (hunk 5 above)
Comment (adamretter) [reply exist-core/src/main/java/org/exist/xquery/CountClause.java]: Done.

File: exist-core/src/main/java/org/exist/xquery/CountClause.java [hunk 6]
```diff
@@ -0,0 +1,211 @@
...
+        while (prev != null) {
+            switch (prev.getType()) {
+                case LET, GROUPBY, FOR -> {
+                    return false;
+                }
+                case ORDERBY -> {
+                    return isDescending(((OrderByClause) prev).getOrderSpecs());
+                }
+            }
+            prev = prev.getPreviousClause();
```
[the remark this answers is not in the document]
Comment (adamretter) [reply exist-core/src/main/java/org/exist/xquery/CountClause.java]: Done.

**Review by reinhapa — APPROVED** [review]

**Comment by adamretter** [comment]
@dizzzz @reinhapa I believe this is now ready to merge please

**Comment by reinhapa** [comment]
@adamretter would you mind changing the this PR to draft mode until it is ready?

**Comment by adamretter** [comment]
> @adamretter would you mind changing the this PR to draft mode until it is ready?

@reinhapa Erm... but... it is ready tobe merged!

**Review by dizzzz — APPROVED** [review]

**Comment by reinhapa** [comment]
@adamretter could you rebase as the tests fail...

**Comment by adamretter** [comment]
> could you rebase as the tests fail...

@reinhapa Done :-)

**Comment by reinhapa** [comment]
@adamretter there are a some simple code smells left, that could be easy solved. Would you mind to address those? (use of new `instanceof` and some usage of `protected` instead of `public` visibility)
